cli/tests/NetworkIO_unittest.cpp

```diff
@@ -0,0 +1,182 @@
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "cli/Flags.hpp"
+#include "cli/LineReaderBuffered.hpp"
+#include "cli/NetworkCliClient.hpp"
+#include "cli/NetworkIO.hpp"
+#include "threading/ConditionVariable.hpp"
+#include "threading/Mutex.hpp"
+#include "threading/Thread.hpp"
+
+#include "glog/logging.h"
+#include "gtest/gtest.h"
+
+using std::unique_ptr;
+
+namespace quickstep {
+
+static std::string const kQueryRequest = "O Captain! My Captain!";
+static std::string const kQueryResponse = "Our fearful trip is done,";
+
+/**
+ * Contains a server instance for testing.
+ */
+class TestNetworkIO {
+ public:
+  TestNetworkIO()
+      : service_(),
+        server_address_("localhost:" + std::to_string(FLAGS_port)),
+        server_(nullptr) {
+    grpc::ServerBuilder builder;
+    builder.AddListeningPort(server_address_, grpc::InsecureServerCredentials());
+    builder.RegisterService(&service_);
+    server_ = builder.BuildAndStart();
+    CHECK(server_) << "Failed to start server";
+    LOG(INFO) << "TestSingleNodeServer listening on " << server_address_;
+  }
+
+  // Gets a message from the Service worker.
+  std::string getSentMessage() {
+    networkio_internal::RequestState& requestState = service_.getRequestState();
+    std::unique_lock<std::mutex> lock(requestState.mutex_);
+    while (!requestState.request_ready_)
+      requestState.condition_.wait(lock);
+
+    EXPECT_EQ(requestState.request_ready_, true);
+    EXPECT_EQ(requestState.response_ready_, false);
+
+    return requestState.request_buffer_;
+  }
+
+  // Sets the response message of the Service worker. Alerts it that the request is ready.
+  void setResponse(std::string response) {
+    networkio_internal::RequestState& requestState = service_.getRequestState();
+    requestState.response_message_.set_query_result(response);
+
+    EXPECT_EQ(requestState.response_ready_, false);
+
+    requestState.responseReady();
+  }
+
+  NetworkCliServiceImpl& getService() {
+    return service_;
+  }
+
+  ~TestNetworkIO() {
+    service_.kill();
+    server_->Shutdown();
+    server_->Wait();
+  }
+
+ private:
+  NetworkCliServiceImpl service_;
+  std::string server_address_;
+  std::unique_ptr<grpc::Server> server_;
+};
+
+/**
+ * Tests that killing the service will cancel requests.
+ */
+TEST(NetworkIOTest, TestShutdown) {
+  TestNetworkIO server;
+
+  server.getService().kill();
+
+  NetworkCliClient client(
+    grpc::CreateChannel("localhost:" + std::to_string(FLAGS_port),
+                        grpc::InsecureChannelCredentials()));
+  QueryRequest request;
+  request.set_query(kQueryRequest);
+  QueryResponse response;
+  Status status = client.SendQuery(request, &response);
+  ASSERT_EQ(status.error_code(), grpc::CANCELLED);
+}
+
+/**
+ * Tests a simple call and response to the Service.
+ */
+TEST(NetworkIOTest, TestNetworkIOCommandInteraction) {
+  NetworkIO networkIO;
+  std::string const query_stmt = kQueryRequest + ";" + kQueryRequest;
```

Comment:
Do you mean `const std::string query_statement`?

Also, why the second query is not ended with `;`?